### PR TITLE
RED-1497 Ruby Client CBP by default

### PR DESCRIPTION
**Pull request by fbvilela**
Tries to make CBP requests when `page` is not specified. If the request raises an error it retries without the cbp specifics

### CODE REVIEW

File: lib/zendesk_api/collection.rb [hunk 1]
```diff
@@ -272,6 +288,7 @@ def prev
         clear_cache
         @options["page"] -= 1
       elsif (@query = @prev_page)
+        @options.page&.delete("after")
```
[the remark this answers is not in the document]
Comment (fbvilela) [reply lib/zendesk_api/collection.rb:292]: the response will have both cursors under `links` and they will be both added to the `next|prev` request parameters if we don't delete the one that does not apply.

File: lib/zendesk_api/collection.rb [hunk 2]
```diff
@@ -188,10 +190,24 @@ def fetch!(reload = false)
...
+      end
+
+      begin
+        @response = get_response(path_query_link)
+      rescue ZendeskAPI::Error::NetworkError => e
+        raise e if intentional_obp_request?
+        @options.per_page = @options_per_page_was
+        @options.page = nil
+        @response = get_response(path_query_link)
+      end
```
Comment (ecoologic) [inline lib/zendesk_api/collection.rb:207]: we're adding more than 10LOC in a method that's already long, we should extract this

File: lib/zendesk_api/collection.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (ecoologic) [reply lib/zendesk_api/collection.rb:292]: I feel like a comment would help

File: lib/zendesk_api/collection.rb [hunk 3]
```diff
@@ -351,17 +365,36 @@ def get_next_page_data(original_response_body)
 
     private
 
+    def cbp_response?(body)
+      body["meta"] && body["links"]
```
Comment (ecoologic) [inline lib/zendesk_api/collection.rb:369]: `!!`

File: spec/core/collection_spec.rb [hunk 4]
```diff
@@ -5,6 +5,58 @@
     ZendeskAPI::Collection.new(client, ZendeskAPI::TestResource)
   end
 
+  # TODO!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+  # TODO: pager, xxx, url
```
Comment (ecoologic) [inline spec/core/collection_spec.rb:9]: will clean up, and I also need to write at least one test for `Helpers#present?`

File: lib/zendesk_api/collection.rb [hunk 5]
```diff
@@ -351,17 +344,64 @@ def get_next_page_data(original_response_body)
...
+        @options.page = { size: (@options_per_page_was || DEFAULT_PAGE_SIZE) }
+      end
+
+      begin
+        @response = get_response(path_query_link)
+      rescue ZendeskAPI::Error::NetworkError => e
+        raise e if intentional_obp_request?
+        @options.per_page = @options_per_page_was
+        @options.page = nil
+        @response = get_response(path_query_link)
```
Comment (ecoologic) [inline lib/zendesk_api/collection.rb:373]: maybe better add a comment here.

File: lib/zendesk_api/collection.rb [hunk 6]
```diff
@@ -351,17 +344,64 @@ def get_next_page_data(original_response_body)
...
+      end
+
+      if path_query_link == "search/export"
```
Comment (ecoologic) [inline lib/zendesk_api/collection.rb:376]: comment why this was left there.

File: lib/zendesk_api/collection.rb [hunk 7]
```diff
@@ -370,13 +415,7 @@ def _all(start_page = @options["page"], bang = false, &block)
...
       while (bang ? fetch! : fetch)
         each do |resource|
-          arguments = [resource, @options["page"] || 1]
-
-          if block.arity >= 0
-            arguments = arguments.take(block.arity)
-          end
-
-          block.call(*arguments)
+          block.call(resource, @options["page"] || 1)
```
Comment (ecoologic) [inline lib/zendesk_api/collection.rb:418]: I can't think of any reason why these two versions should behave differently -> simplifying.

**Review by JinHuangMelb — APPROVED** [review]
LGTM

**Pull request merged**